src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,646 @@
+                } else {
+                    documentTermFrequency.insert(std::make_pair(std::make_tuple(iterStart, iterEnd), 1));
+                }
+            }
+        );
+
+        std::float_t normVal = 0.0f;
+        std::vector<std::tuple<std::uint32_t, std::float_t>> results;
+        for (auto const & wordIteratorPair : documentTermFrequency) {
+            std::string const word(std::string(std::get<0>(wordIteratorPair.first), std::get<1>(wordIteratorPair.first)));
+
+            IndexMap::const_iterator const      labelIter(Labels.find(word));
+
+            if (labelIter != Labels.end()) {
+
+                double tf;
+                double idf;
+
+                //calculate tf(term frequency) which measures how frequently a term occurs in a document.
+                //Since every document is different in length, it is possible that a term would appear much more times
+                //in long documents than shorter ones. Thus, the term frequency is often divided by the document length
+                //(aka. the total number of terms in the document) as a way of normalization:
+                //TF(t) = (Number of times term t appears in a document) / (Total number of terms in the document)
+                //source:http://www.tfidf.com/
+                if ((TfidfParameters & TfidfPolicy::Binary) == TfidfPolicy::Binary) {
+                    tf = 1.0;
+                } else if (!((TfidfParameters & TfidfPolicy::SublinearTf) == TfidfPolicy::SublinearTf)) {
+                    tf = wordIteratorPair.second;
+                } else {
+                    tf = 1.0 + std::log(wordIteratorPair.second);
+                }
+
+                //calculate idf(inverse document frequency) which measures how important a term is. While computing TF,
+                //all terms are considered equally important. However it is known that certain terms, such as "is", "of",
+                //and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent
+                //terms while scale up the rare ones, by computing the following:
+                //IDF(t) = log_e(Total number of documents / Number of documents with term t in it).
+                //source:http://www.tfidf.com/
+                if (!((TfidfParameters & TfidfPolicy::UseIdf) == TfidfPolicy::UseIdf)) {
+                    idf = 1.0;
+                } else if ((TfidfParameters & TfidfPolicy::SmoothIdf) == TfidfPolicy::SmoothIdf) {
+                    idf = 1.0 + std::log((1 + TotalNumsDocuments) / (1.0 + DocumentFreq.at(word)));
+                } else {
+                    idf = 1.0 + std::log((1 + TotalNumsDocuments) / (0.0 + DocumentFreq.at(word)));
+                }
+
+                //calculate tfidf (tfidf = tf * idf)
+                std::float_t tfidf = static_cast<std::float_t>(tf * idf);
+
+                //calculate normVal
+                if(Norm == NormMethod::L1) {
+                    normVal += tfidf;
```

Comment:
Since abs() is removed here, we should add an assert(tfidf >=0 ) and normVal(>= 0)

Reply:
added